### PR TITLE
Setting to strip EXIF metadata from JPEG uploads

### LINKED ISSUE #4426: Options to strip EXIF metadata from media uploads.
A really useful privacy feature would be to be able to strip off any metadata from images (or audio or video) before uploading them to the server.  This could be done at the server itself, but then the server would be able to sniff it and it wouldn't work with E2E uploads which are probably the most important ones anyway.

So instead we should probably implement it as an option for the client to mangle the image before uploading to strip it out.

### PR DESCRIPTION
**Pull request by t3chguy**
This strips EXIF metadata from uploaded JPEGs by default, thus avoiding privacy sensitive metadata such as GPS and timestamps being leaked over Matrix.

Fixes https://github.com/vector-im/riot-web/issues/4426

### CODE REVIEW

**Comment by t3chguy** [comment]
Leads to broken thumbnails:
image

```
{
  "origin_server_ts": 1502833036623,
  "sender": "@abc123abc:matrix.org",
  "event_id": "$15028330362519985ISVKc:matrix.org",
  "unsigned": {
    "age": 7299
  },
  "content": {
    "body": "12382975864_09e6e069e7_o.jpg",
    "info": {
      "mimetype": "image/jpeg",
      "thumbnail_info": {
        "mimetype": "image/jpeg",
        "h": 600,
        "w": 800,
        "size": 205399
      },
      "h": 2976,
      "thumbnail_url": "mxc://matrix.org/nNjRtwWgHjjLCvXCKuubQAfx",
      "w": 3968,
      "size": 6327505
    },
    "msgtype": "m.image",
    "url": "mxc://matrix.org/KPImtwzFwwJDXVSthmbyedRh"
  },
  "type": "m.room.message",
  "room_id": "!UInSXSErRtlgialnsk:riot.ovh"
}
```

**Comment by t3chguy** [comment]
/me is unsure what is going on :/

**Comment by lukebarnard1** [comment]
@ara4n any update on this?

**Comment by ara4n** [comment]
not from me; it didn't work for @t3chguy and we haven't yet debugged it.

**Comment by jryans** [comment]
Thanks for making this contribution a while back. Since the code base has changed since this was opened, it no longer applies cleanly, and I don't think there's a need to keep it open in this state, as we can always find the code here again if needed.

It looks like we should probably file an issue (if there isn't one already) to track the problem being solved here.

**Comment by ara4n** [comment]
https://github.com/vector-im/riot-web/issues/4426 is the bug.

**Comment by ara4n** [comment]
So the reason why the thumbnails weren't working is that PIL in synapse silently chokes on thumbnailing the exif-stripped image, and so the server-side generated thumbnails 404.

**Comment by ara4n** [comment]
empirically this seems to work now. tested in e2ee & non-e2ee rooms, with small & massive images.

It's worth noting that you have to try quite hard to make Riot/Web use server-generated thumbnails these days since https://github.com/matrix-org/matrix-react-sdk/pull/2439; images have to be bigger than 1MB and larger than 1600x1200 (on a retina display) before riot-web will actually ask synapse for a thumbnail.

File: src/settings/Settings.js [hunk 1]
```diff
@@ -229,6 +229,11 @@ export const SETTINGS = {
         displayName: _td('Enable automatic language detection for syntax highlighting'),
         default: false,
     },
+    "stripJpegMetadata": {
```
Comment (dbkr) [inline src/settings/Settings.js:232]: Would suggest probably `stripUploadMetadata` or maybe `stripImageMetadata`: if we start stripping metadata from other image file types we'll want to use the same setting (the description should continue to say JPEG whilst that's all it does).

File: src/ContentMessages.js [hunk 2]
```diff
@@ -258,13 +323,29 @@ function readFileAsArrayBuffer(file) {
  *  If the file is encrypted then the object will have a "file" key.
  */
 function uploadFile(matrixClient, roomId, file, progressHandler) {
+    let readDataPromise;
+
+    // strip metadata where we can (n.b. we don't want to strip colorspace metadata)
+    if (file.type === 'image/jpeg' &&
+        SettingsStore.getValueAt(SettingLevel.ACCOUNT, 'stripImageMetadata')) {
```
Comment (dbkr) [inline src/ContentMessages.js:330]: Why is this specifically getting the value at account level?

File: src/ContentMessages.js [hunk 3]
```diff
@@ -258,13 +323,29 @@ function readFileAsArrayBuffer(file) {
...
+        readDataPromise = readFileAsArrayBuffer(file).then(function(data) {
+            // fix up arraybuffer
+            data = stripJpegMetadata(data);
+            return Promise.resolve(data);
```
Comment (dbkr) [inline src/ContentMessages.js:334]: shouldn't this just be `return data`?

**Review by dbkr — COMMENTED** [review]
Also the two FIXMEs look like we should establish whether they do indeed break things in the way that they say they might.

**Comment by jryans** [comment]
Thanks for making this contribution a while back. Since the code base has changed since this was opened, it no longer applies cleanly, and I don't think there's a need to keep it open in this state, as we can always find the code here again if needed. If you are still interested in pursuing this feature, please discuss with us in #element-dev:matrix.org to find a good approach forward.

**Comment by t3chguy** [comment]
One deficiency with this PR is that it strips the `Orientation` EXIF which will cause images to rotate

image

top is with EXIF stripping
bottom is without

image
image

**Comment by t3chguy** [comment]
So I don't see a sane way to retain the Orientation IFD0 - as for the `colorimetry` we definitely strip `colourSpace` and all the chromacities stuff, all residing in `APP1` along with `IFD0`

**Comment by t3chguy** [comment]
So whilst this won't work ideally in all cases, I think its good enough for 90% of cases.

**Comment by ara4n** [comment]
ugh. i failed to think through that this would break orientation and colorimetry. can we exclude those from being stripped? last thing we need is to make orientation even more flakey.

**Comment by t3chguy** [comment]
> can we exclude those from being stripped?

Not easily, would be saner to load re-export out of a canvas IMO

**Comment by Palid** [comment]
@t3chguy this PR is here for a while now, does it still apply? If yes, wouldn't it make sense to make a new issue for it so we can properly prioritise it take care of it later?
It seems that you've figured out some problems while working at it, I think it'd be a good idea to add those in the issue as the definition of done.

**Comment by t3chguy** [comment]
> does it still apply? If yes, wouldn't it make sense to make a new issue for it so we can properly prioritise it take care of it later?

@Palid what's wrong with the existing issue https://github.com/vector-im/element-web/issues/4426 ?

**Comment by Palid** [comment]
> > does it still apply? If yes, wouldn't it make sense to make a new issue for it so we can properly prioritise it take care of it later?
> 
> @Palid what's wrong with the existing issue https://github.com/vector-im/element-web/issues/4426 ?

Like a lot of old issues it has spread thin in the details over the years and it's hard to see if it's still relevant. I think this issue is still relevant, but it made more sense to ping you. I think that  details of the issue should be probably updated with some newer informations that we already have after initial implementations.

**Comment by olmari** [comment]
I might be late on this, but...

While I couldn't _definitively_ be against _option_ to strip metadata form cerrtain files if _user chooses to_, I'm totally against matrix (umbrella) touching anything in uploaded media _by default_. That is so big can of worms that will never stop even if you could figure out to do prefect stripping for every file format you accept to touch... And not to say you even can't do it really selectively for images.. (rotation, colorspace, etc data)...

https://github.com/vector-im/element-web/issues/4426#issuecomment-886918321 is still spot on comment on the thing.. Specificly: "private and secure is not the same as anonymous", those who don't want to exif data to "leak", they shoudn't send images with exif data in first place... It can't be matrix task to decide for user, at the least not by default...

**Comment by ShadowJonathan** [comment]
Existing social media platforms (Twitter, discord, telegram) already do this, they already strip potentially user-compromising data out of media, element would only be following the trend.

I do agree that there'd be an option to disable it, but I think that element would be making a right choice for (at least) removing GPS data from images wherever possible.

**Pull request closed without merging**